src/tensor_ops/ops.rs

```diff
@@ -91,3 +91,23 @@ pub(crate) fn try_binary_op<
     });
     Ok(out.put_tape(tape))
 }
+
+/// computes the strides needed to store the result of a binary tensor operation given the strides
+/// of the input tensors
+// TODO: put this somewhere better
+pub(crate) fn merge_strides<S: Shape>(
+    shape: S,
+    mut strides1: S::Concrete,
+    strides2: S::Concrete,
+) -> (usize, S::Concrete) {
+    let mut prod = 1;
+    let dims = shape.concrete();
+
+    for i in (0..S::NUM_DIMS).rev() {
+        if strides1[i] > 0 || strides2[i] > 0 {
+            strides1[i] = prod;
+            prod *= dims[i];
+        }
+    }
+    (prod, strides1)
```

Comment:
So this function is essentially saying "if both array's stride is 0 in this dimension, keep it at 0"? E.g. if I take two tensors with dim 0 broadcasted, then the resulting tensor will also have dim 0 broadcasted.

Reply:
Yes, each stride of the output tensor is only broadcasted if that dimension is broadcasted for both of the input tensors to ensure that the output tensor can hold all unique resulting values.

Reply:
Gotcha, that makes sense. I like this change, but it does make what the strides will be a little harder to reason about.

Let's defer until after cuda stuff is more stable. Additionally, it doesn't look like pytorch does this optimization, and I would like to investigate why they didn't (it seems to be a pretty straightforward one):
```python
>>> import torch
>>> a = torch.randn(2, 3)
>>> b = torch.randn(2, 3)
>>> c = a.broadcast_to(5, 2, 3) + b.broadcast_to(5, 2, 3)
>>> c.stride()
(6, 3, 1)
```